automap/src/comm_layer/igdp.rs

```diff
@@ -0,0 +1,1687 @@
+    }
+
+    #[test]
+    fn ensure_gateway_handles_missing_gateway() {
+        let gateway_factory =
+            GatewayFactoryMock::new().make_result(Err(SearchError::InvalidResponse));
+        let mut subject = IgdpTransactor::new();
+        subject.gateway_factory = Box::new(gateway_factory);
+
+        let result = subject.ensure_gateway();
+
+        assert_eq!(result, Err(AutomapError::CantFindDefaultGateway));
+    }
+
+    #[test]
+    fn thread_guts_iteration_handles_missing_public_ip() {
+        init_test_logging();
+        let new_public_ip = Ipv4Addr::from_str("4.3.2.1").unwrap();
+        let gateway = GatewayWrapperMock::new().get_external_ip_result(Ok(new_public_ip));
+        let inner_arc = Arc::new(Mutex::new(IgdpTransactorInner {
+            gateway_opt: Some(Box::new(gateway)),
+            housekeeping_commander_opt: None,
+            public_ip_opt: None,
+            mapping_adder: Box::new(MappingAdderMock::new()),
+            logger: Logger::new("test"),
+        }));
+        let change_log_arc = Arc::new(Mutex::new(vec![]));
+        let change_log_inner = change_log_arc.clone();
+        let change_handler: ChangeHandler =
+            Box::new(move |change| change_log_inner.lock().unwrap().push(change));
+
+        let result = IgdpTransactor::thread_guts_iteration(
+            &change_handler,
+            &inner_arc,
+            &mut Instant::now(),
+            &Some(MappingConfig {
+                hole_port: 6666,
+                next_lifetime: Duration::from_secs(0),
+                remap_interval: Duration::from_millis(1000),
+            }),
+        );
+
+        assert!(result);
+        let change_log = change_log_arc.lock().unwrap();
+        assert_eq!(
+            *change_log,
+            vec![AutomapChange::NewIp(IpAddr::V4(new_public_ip))]
+        );
+        TestLogHandler::new().exists_log_containing(
+            "WARN: test: Housekeeper was started before retrieving public IP",
+        );
+    }
+
+    #[test]
+    fn thread_guts_iteration_handles_missing_mapping_config() {
+        let new_public_ip = Ipv4Addr::from_str("4.3.2.1").unwrap();
+        let gateway = GatewayWrapperMock::new().get_external_ip_result(Ok(new_public_ip));
+        let inner_arc = Arc::new(Mutex::new(IgdpTransactorInner {
+            gateway_opt: Some(Box::new(gateway)),
+            housekeeping_commander_opt: None,
+            public_ip_opt: Some(new_public_ip),
+            mapping_adder: Box::new(MappingAdderMock::new()),
+            logger: Logger::new("thread_guts_iteration_handles_missing_mapping_config"),
+        }));
+        let change_handler: ChangeHandler = Box::new(move |_| panic!("Shouldn't be called"));
+
+        let result = IgdpTransactor::thread_guts_iteration(
+            &change_handler,
+            &inner_arc,
+            &mut Instant::now().sub(Duration::from_secs(1)),
+            &None,
+        );
+
+        assert!(result);
```

Comment:
I'd think of adding a debug!() logger to referring to a void after the mapping config. I know we don't assert on them but still it could be handy in the future.